phoenix-core/src/main/java/org/apache/phoenix/monitoring/TableMetricsManager.java

```diff
@@ -0,0 +1,287 @@
+                            }
+
+                        } else {
+                            LOGGER.error(
+                                    "Phoenix table level metrics publisher className cannot be null");
+                        }
+
+                    }
+                }
+            }
+        }
+        return tableMetricsManager;
+    }
+
+    /**
+     * This function is provided as hook to publish the tableLevel Metrics to
+     * LocalStore(tablePhoenixMapping).
+     *
+     * @param map of tableName to pair of (MetricType, Metric Value)
+     */
+    public void pushMetricsFromConnInstance(Map<String, Map<MetricType, Long>> map) {
+
+        if (map == null) {
+            LOGGER.debug("Phoenix table level metrics input map cannott be null");
+            return;
+        }
+
+        long startTime = EnvironmentEdgeManager.currentTime();
+        for (Map.Entry<String, Map<MetricType, Long>> tableEntry : map.entrySet()) {
+            for (Map.Entry<MetricType, Long> metricEntry : tableEntry.getValue().entrySet()) {
+                updateMetrics(tableEntry.getKey(), metricEntry.getKey(), metricEntry.getValue());
+            }
+        }
+
+        LOGGER.debug(String.format(
+                "Phoenix table level metrics completed updating metrics from conn instance, timetaken:\t%d",
+                +EnvironmentEdgeManager.currentTime() - startTime));
+    }
+
+    /**
+     * This function will be used to add individual MetricType to LocalStore.
+     *
+     * @param tableName
+     * @param type
+     * @param value
+     */
+    public void updateMetrics(String tableName, MetricType type, long value) {
+
+        long startTime = EnvironmentEdgeManager.currentTime();
+
+        TableClientMetrics tInstance = getTableClientMetrics(tableName);
+        if (tInstance == null) {
+            LOGGER.debug("Table level client metrics are disabled for table: " + tableName);
+            return;
+        }
+        tInstance.changeMetricValue(type, value);
+
+        LOGGER.debug(String.format("Phoenix table level metrics completed updating metric"
+                        + " %s to value %s, timetaken = %s", type, value,
+                EnvironmentEdgeManager.currentTime() - startTime));
+    }
+
+    /**
+     * Get Table specific metrics object and create if not initialized(thread safe)
+     *
+     * @param tableName
+     * @return TableClientMetrics object
+     */
+    private TableClientMetrics getTableClientMetrics(String tableName) {
+
+        if (Strings.isNullOrEmpty(tableName)) {
+            LOGGER.debug("Phoenix Table metrics TableName cannot be null or empty");
+            return null;
+        }
+
+        if (!allowedListOfTableNames.isEmpty() && !allowedListOfTableNames.contains(tableName)) {
+            return null;
+        }
+
+        TableClientMetrics tInstance;
+        tInstance = tableClientMetricsMapping.get(tableName);
+        if (tInstance == null) {
```

Comment:
similar comments here about correcting the double-checked locking

Reply:
Modified the implementation.